### PR TITLE
fix(claude-code): synthesize root SKILL.md Read span + inject matching tool_call into LLM output for /skill (AGE-1126)

### PR DESCRIPTION
**Pull request by rangemer333-cell**
## 问题

`/skill` 命令加载技能时，根 `SKILL.md` 的 Read 未在 trace 中生成对应的 Read span——CLI 通过内部 `Skill` 工具隐式读取根 `SKILL.md`，不经过显式 `Read` 工具调用，导致 owner step 的 LLM output 里只有 `Skill` tool_call，缺失根 `SKILL.md` 的 Read span 与 LLM 背书，`validate-trace` 报 `semantic.tool_matches_llm_output` ERROR。

## 方案
1. **合成根 SKILL.md Read span**：在解析 transcript 时，为 `/skill` 加载的根 `SKILL.md` 合成一个 Read span，挂到触发 `/skill` 的 owner step（step1）之下，call/result 使用同一 id。
2. **向 owner LLM output_messages 注入同 id tool_call**：向 owner step1 的 LLM span `gen_ai.output.messages` 注入一条 name=Read、call id 与合成 span 一致的 tool_call，使合成的 Read span 获得 LLM 背书，消除 `semantic.tool_matches_llm_output` ERROR。原有 `Skill` tool_call 并存，不被覆盖。

### 去重键

以 (session, 根 SKILL.md 路径, owner step) 为去重键，确保根 SKILL.md 在 history + OTLP 两侧各恰好 1 个 Read span，避免重复合成。

## 影响范围

- **references / 普通文件不受影响**：技能内 `references/guide.md` 等的真实 Read 保留原样；Bash/Write 等普通工具调用不受影响，无误删。
- 仅针对 `/skill` 根 SKILL.md 隐式读取这一路径做合成。

### 单测
- `tests/unit/hooks/claude-code/skill-read.test.mjs` 新增覆盖合成 span + 注入 tool_call + 去重 + references 保留 + 普通文件不误删。
- **单测 55/55 通过**。

### CP5 真实 E2E（0 ERROR）
- 目标 Agent CLI：`claude` 2.1.175 (Claude Code)，容器内实时鉴权真实 live 多轮 ReAct 对话。
- 触发：`/skill cp5-skill-probe` + Read guide + 2×Bash + Write + 总结，产出真实 trace。
- 复杂度：**5 STEP / 5 TOOL**（Skill×1、Read×2、Bash×2）。
- `validate-trace`：**PASS，0 ERROR**（上轮 `semantic.tool_matches_llm_output` 已消除，pass 数 32→33；WARN 全为可选 SHOULD 属性，非阻塞）。
- 背书断言：合成根 SKILL.md Read 的 tool_call 已出现在 owner step1 LLM span `gen_ai.output.messages`（同 call id / name=Read），原 `Skill` tool_call 并存。
- `gen_ai.input/output.messages` 全 LLM span 非空。

## 关联

AGE-1126

### CODE REVIEW

**Comment by rangemer333-cell** [comment]
## Step C 加固 commit (`7153fac`)

在原 commit `7a22343` 基础上追加一次加固，回应 architect review（#1+#2 定为 P1，本轮落地；#3/#4 经 architect 裁定 DEFER 下一轮）。本轮仅动 `transcript-parser.mjs` + 单测，`claude-code-hook-processor.mjs` 未改。

### #1 解耦 skill 注入识别（加法式，绝不更窄）
- skill 注入的识别改为**结构信号 OR 前缀信号**的加法式判定：任一命中即识别为 `/skill` 注入，避免单一信号（前缀/结构）在上游格式漂移时漏判。
- 严格加法：新逻辑只会**扩大或维持**识别范围，绝不比原判定更窄——已识别的场景不会因加固而漏掉。

### #2 `skill_root_parse_miss` 漂移可观测
- 当识别为 skill 注入但根 SKILL.md 解析未命中时，发出 `skill_root_parse_miss` 漂移信号，使上游 transcript 格式变化可被观测，而非静默降级。

### DEFER（architect 裁定，下一轮）
- **#3 兜底**：解析全 miss 时的兜底策略 —— DEFER。
- **#4 合成 Read result 内容**：合成 Read span 的 result 内容完善 —— DEFER。

### 验收
- 单测 **58/58** ✅（新增 3 例覆盖加法式识别 + 漂移信号）。
- Step C 回归 E2E **0 ERROR** ✅：根 Read span=1 / 背书 / references / 普通文件与 CP5 逐字节等价，无回归。
- architect **APPROVED**。

关联 AGE-1126。

**Review by ralf0131 — APPROVED** [review]
LGTM. Well-designed synthesized Read span for /skill commands.

- `normalizeSkillPath()` correctly avoids toLowerCase/realpathSync for cross-platform safety
- `deterministicSkillReadId()` ensures idempotent call IDs
- `injectSynthesizedToolCall()` properly places the synthesized tool_call in LLM output for ARMS semantic validation
- Drift detection via logHookError is a good defensive measure

Test coverage (506 lines) covers all edge cases.

---
*Automated review by github-manager-bot*

**Comment by rangemer333-cell** [comment]
## 不变量护栏单测 (`f4849c7`) — test-only

追加 2 条不变量护栏单测，**生产代码零改动**，锁定跨窗口解析的关键前提，防未来重构静默丢根 Read span：

- **同窗 Skill + meta**：`Skill` tool_use 与 skill meta 注入必须在**同一解析窗口**内被解析到（若被拆到不同窗口，根 SKILL.md Read 的合成与背书就会断链）。
- **offset 仅 Stop 推进 / 无 mid-turn hook**：解析 offset 只由 Stop 事件推进，当前不存在 mid-turn hook。一旦未来误加 mid-turn hook 打破此前提，该测试立即变红——把「跨窗口静默丢根 Read」这一隐患转成编译期可见的失败。

单测 **60/60** 绿。无生产代码变更，无需重部署/E2E。关联 AGE-1126。

**Comment by fangxiu-wf** [comment]
感谢补充 `/skill` 根 `SKILL.md` 的采集能力，以及后续对 parser 漂移和生命周期不变量的加固。我基于最新 HEAD `f4849c7` 重新检查了生产代码、全部 PR 讨论和测试；本地 Claude Code 相关用例 **60/60 通过**。此外，我补了 3 个针对去重、时间和 checkpoint 边界的复现场景。

当前结论：**Request changes，暂不建议直接合入。** 正常单次 `/skill` 路径工作正常，但仍有两个会影响 Trace 真实性/完整性的 P1。

### [P1] 合成 Read 被写成了模型真实输出

[`injectSynthesizedToolCall`](https://github.com/alibaba/loongsuite-pilot/blob/f4849c741e9499b1c87223c17bd859bba818118a/assets/hooks/claude-code-hook-processor.mjs#L228-L244) 会把 Claude 实际未输出的 `Read` tool_call 注入 owner LLM 的 `gen_ai.output.messages`；调用点在 [L740-L747](https://github.com/alibaba/loongsuite-pilot/blob/f4849c741e9499b1c87223c17bd859bba818118a/assets/hooks/claude-code-hook-processor.mjs#L740-L747)。实际 transcript 中模型只声明了 `Skill`，根文件读取发生在 Claude Code 的内部 Skill 执行阶段。

因此现在的 Trace 会错误地表示“模型主动调用了 Read”。这确实能让 `semantic.tool_matches_llm_output` 变绿，但通过改变模型输出记录来满足 validator，会影响审计、行为分析和后续训练数据的真实性。仓库 Schema 和 OpenTelemetry 对 `gen_ai.output.messages` 的定义都是模型返回的消息。

建议：

- 将该 Read 明确建模为 `derived/synthetic` 的内部工具执行，并添加稳定的 provenance 属性；
- 调整 `semantic.tool_matches_llm_output`，让它对明确标记的内部派生工具采用单独校验逻辑，或扩展 converter 表达 Skill → Read 的内部父子关系；
- 不要在 LLM span 的真实 output 中伪造一个模型没有返回的 tool_call。Cursor 现有实现采用类似方式，可以作为兼容背景，但不能替代对数据语义的明确约定。

### [P1] 实际去重范围不是 PR 声明的 `(session, path, owner step)`

当前实现使用 turn 级的 [`realReadPaths`](https://github.com/alibaba/loongsuite-pilot/blob/f4849c741e9499b1c87223c17bd859bba818118a/assets/hooks/claude-code-hook-processor.mjs#L627-L638) 和 `synthesizedSkillRoots`，确定性 ID 也只有 `client + turnId + rootPath`；[合成判断](https://github.com/alibaba/loongsuite-pilot/blob/f4849c741e9499b1c87223c17bd859bba818118a/assets/hooks/claude-code-hook-processor.mjs#L699-L713) 没有 owner step。

我构造了“同一 turn、两个不同 LLM owner step 连续调用同一个 Skill”的场景，稳定得到：

- `Skill tool span = 2`
- `Read tool span = 1`

第二次 Skill 所隐含的根文件读取被 `synthesizedSkillRoots` 抑制。此外，只要同一 turn 的任意其他 step 出现相同路径的真实 Read，也会抑制 Skill owner step 的合成 Read。这与 PR 描述的 owner-step 去重键不一致，也会把两个真实发生的读取折叠成一个。

建议按 `owner.stepId + normalizedRootPath`（或直接按 Skill tool call ID）关联真实/合成记录并生成 ID；如果产品契约实际就是“每 turn 每路径最多一次”，则需要修改 PR 描述和测试，并明确接受重复 Skill 调用被折叠的行为。

### [P2] 缺少 Skill result 时间时会生成零时长 Read span

代码使用：

```js
const readResultTs = timestamps.result || timestamps.call;
```

随后分别把 call/result 写在 `timestamps.call` 和 `readResultTs`，[见 L711-L737](https://github.com/alibaba/loongsuite-pilot/blob/f4849c741e9499b1c87223c17bd859bba818118a/assets/hooks/claude-code-hook-processor.mjs#L711-L737)。当 result timestamp 缺失时，两者完全相同；我已稳定复现。converter 会按这两个时间结束 TOOL span，最终触发 `time.non_zero_duration`。

建议保留 isMeta 注入记录的 timestamp，优先作为合成 Read 的真实结束时间；没有可用时间时至少保证 result 严格晚于 call，并增加 `result.time_unix_nano > call.time_unix_nano` 的回归断言，而不只是检查 timestamp 非 `0`。

### [P2] 空字符串 result 会混淆“真实空结果”和“未采集到结果”

当前合成结果固定写入：

```js
'gen_ai.tool.call.result': '',
'tool.result.status': 'success'
```

消费者无法判断这是读取了空文件，还是合成逻辑没有记录结果。`gen_ai.tool.call.result` 本身是可选字段；若暂不采集正文，建议省略 result 属性并增加 synthetic/provenance 标记，而不是记录一个看似真实的空结果。作者评论中已将结果内容完善列为 DEFER，我建议至少先避免写入有歧义的空值。

### 已排除项

我也复核了“Skill 与 isMeta 被拆到两个 parse window”这一风险：强行在二者之间手工执行 Stop 的确会导致最终 `Read=0`，但当前 Pilot 只在 Stop 推进 offset，Claude Code 的 Stop 生命周期也位于响应结束。因此在现有正常生命周期下，我不把它列为合入阻断项；`f4849c7` 新增的不变量测试对未来引入 mid-turn hook 的回归风险也有帮助。

总体上，当前实现对单次标准 `/skill` 场景覆盖充分，references 和普通 Read 的既有行为没有发现回归。建议先解决上面两个 P1，并补充 owner-step 重复 Skill 与非零 duration 的端到端断言后再合入。

File: assets/hooks/claude-code-hook-processor.mjs [hunk 1]
```diff
@@ -628,6 +695,58 @@ function buildTurnRecords(turn, turnIndex, sessionId, prevHash, userId, turnStop
         }
         records.push(resultRecord);
       }
+
+      // 显式 /skill: 保证本 turn 根 SKILL.md 恰好展示一次 Read span。
+      // 去重键 = client + turnId + 归一化根路径。若本 turn 已有真实根 Read 则不合成
+      // (真实事件优先);否则合成一条同 call-id 的 Read tool.call+tool.result,挂 Skill owner step。
+      if (toolName === 'Skill' && skillRootByToolId) {
```
Comment (linrunqi08) [inline assets/hooks/claude-code-hook-processor.mjs:702]: **[Medium]** 跨 parse-window 边界会漏合成根 SKILL.md Read span。

合成只在此处 `toolName === 'Skill'` 分支触发（依赖当前 window 的 `declaredToolIds`），而根路径只在后续 `isMeta` 注入里出现（transcript-parser.mjs:246）。若二者被 window 边界切开（`Stop`/export 落在 Skill 块与其注入之间，或 `MAX_TRANSCRIPT_BYTES` 50MB 尾截断切开）：window N 发裸 `Skill` span 无合成；window N+1 拿到路径但本 window 无 Skill toolId → 仍不合成。

**影响:** 该 `/skill` turn 的根 SKILL.md Read 永不产出，本 PR 要修复的 `semantic.tool_matches_llm_output` 对该 turn 静默失效。此「漏合成」分支未被任何断言覆盖（场景 7 只覆盖「重复合成」变体）。现实概率低（正常单 turn 内 Skill 调用/结果/注入同 window）。

**建议:** 至少在生产代码注释显式标注该「漏合成」边界（当前仅 test 注释了重复变体）；若要闭合，可将未解析的 Skill toolId + 待定根路径持久化到 state，让后续 window 补合成。

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary

| Severity | Count |
|----------|-------|
| Critical | 0 |
| High     | 0 |
| Medium   | 1 |
| Low      | 4 |

本轮评审 head: `f4849c74`（生产代码与 `7153facf` 完全一致，`f4849c74` 仅新增测试不变量护栏）· base(merge-base): `e4c78901` · 本地 `vitest run tests/unit/hooks/claude-code/` **58/58 通过**（含新增 skill-read 15 例），无回归。

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS |
| 状态恢复正确性 | PASS |

- 未引入定时器/句柄/子进程；新增 Map/Set 均为 parse-window / turn 级，随作用域 GC。
- 纯同步转换，无锁/无新增 await；`realReadPaths` 每 turn 构建一次（`size>0` 才触发），无 O(n²)。
- fail-open 完备，消息 hash 基于 input 不受 output 注入影响。唯一状态缺口为下述 Medium「跨 window 漏合成」，低概率且不影响宿主。

### 总体结论

**未发现阻断合入问题（Critical/High = 0），建议可合入。** 变更为 `/skill` 隐式读取根 SKILL.md 补齐 Read span 并满足 ARMS `semantic.tool_matches_llm_output`，三层去重（真实优先 / turn 内幂等 / 确定性 call id）设计严谨，对普通文件 / references / 非 skill 路径零副作用。

主要风险点（1 个 Medium，**评审中已被作者 commit `f4849c74` 基本闭合**）：`Skill` tool_use 与携带根路径的 `isMeta` 注入若被 parse-window 边界切开，该 turn 合成 Read 会漏产、ARMS 修复对该 turn 静默失效。作者新增的「同窗不变量」测试从源码层静态钉死：仅注册 Stop 类 hook、`transcript_offset` 仅在 Stop 导出后推进、transcript 仅在 Stop 解析一次 → 解析窗口恒对齐 turn 末尾，Skill 与其注入必落同窗，该「漏合成」在现状模型下不可达；且一旦未来误加 mid-turn hook 使其可达，护栏立即变红。评价：这正是行内建议的「显式标注边界」的更强落地（静态守卫 > 注释）。其余 4 项 Low 详见行内评论与报告。

### Highlights（正向实践）

- 真实优先 + turn 内幂等 + 确定性 call id 三层去重清晰，call/result 复用同 span_id/id → OTLP 稳定汇聚一个 span。
- fail-open 严格保持；`skill_root_parse_miss` 经「结构确认 + 前缀失配」双 gate，不误报、不刷屏，把协议漂移的静默回归转成可检测信号。
- 消息 hash 无副作用（注入只改 output，hash 基于 input）。
- 测试扎实：15 例覆盖 parser 捕获 / 尾斜杠 / 跨 window / 漂移打点 + 6 端到端场景，并显式记录已知边界。

---
*评审报告详见: `code-review/pr-194/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

File: assets/hooks/claude-code-hook-processor.mjs (hunk 1 above)
Comment (linrunqi08) [reply assets/hooks/claude-code-hook-processor.mjs:702]: 补充：注意到评审期间的 commit `f4849c74` 已新增「同窗不变量」测试护栏（静态钉死仅 Stop 类 hook、offset 仅 Stop 推进、transcript 仅 Stop 解析一次），证明该「漏合成」分支在现状模型下不可达，并对未来误加 mid-turn hook 提供回归护栏。此项已基本闭合，本条保留作背景说明。

**Review by ralf0131 — APPROVED** [review]
### Summary
LGTM — This PR correctly synthesizes root SKILL.md Read spans and injects matching tool_calls into LLM output for `/skill` commands, fixing the `semantic.tool_matches_llm_output` validation error. The implementation:

1. **Well-designed synthesis** — `normalizeSkillPath()` uses pure lexical normalization (no realpathSync/toLowerCase pitfalls), `deterministicSkillReadId()` ensures idempotency via hash-based id
2. **Correct injection** — `injectSynthesizedToolCall()` properly injects into LLM output.messages without overwriting existing Skill tool_call
3. **Robust deduplication** — (client, turnId, normalizedRootPath) key prevents duplicate synthesis
4. **Drift observability** — `skill_root_parse_miss` error logging with structural gate prevents false positives while detecting prefix drift
5. **Comprehensive tests** — 7 end-to-end scenarios + cross-window parsing + prefix drift detection + invariant tests document known boundaries

The cross-hook-run delayed Read boundary (场景7) is well-documented as a non-P0 issue.

---
*Automated review by github-manager-bot*

**Review by linrunqi08 — APPROVED** [review]
